tests/benchmark/compute/instruction/test_system.py

```diff
@@ -247,6 +247,55 @@ def test_xcall(
     )
 
 
+def test_contract_calling_many_addresses(
+    benchmark_test: BenchmarkTestFiller,
+    pre: Alloc,
+    fork: Fork,
+) -> None:
+    """
+    Benchmark a contract that calls many addresses.
+
+    Creates a contract that generates addresses dynamically using KECCAK256
+    and calls each one with 1 wei.
+    Stop when the gas is less than the loop cost.
+    """
+    gas_costs = fork.gas_costs()
+    loop_cost = (
+        gas_costs.G_VERY_LOW * 3  # ~MSTOREs+ADDs
+        + gas_costs.G_COLD_ACCOUNT_ACCESS  # CALL cost
+        + gas_costs.G_CALL_VALUE  # Value transfer cost
+        + gas_costs.G_LOW * 3  # ~Gluing opcodes
+    )
+    # Build contract code that generates addresses and calls them
+    # Pattern:
+    # 1. Store Keccak256(NUMBER) in memory[0] (32 bytes)
+    # 2. JUMPDEST (loop start)
+    # 4. CALL address from memory[0] (first 20 bytes) with 1 wei
+    # 5. Update memory[0] with ADD(memory[0], 1)
+    # 6. JUMP back to JUMPDEST
+    setup = Op.MSTORE(0, Op.NUMBER())  # Initialize with block number
+    destination = Op.JUMPDEST()
+    attack_block = Op.MSTORE(
+        0,
+        Op.ADD(
+            1,
+            Op.CALL(
+                address=Op.MLOAD(0),  # Load address from memory
+                value=1,  # Send 1 wei
+            ),
+        ),
+    )
+    attack_block += Op.JUMPI(Op.GT(Op.GAS, loop_cost), destination)
```

Comment:
```suggestion
    attack_block += Op.JUMPI(Op.GT(Op.GAS, loop_cost), len(setup))
```

The program count provided should be an index instead of an opcode.